src/ui_nbgl.c

```diff
@@ -0,0 +1,160 @@
+                .text = "Allow contract data",
+                .subText = NULL,
+                .token = CONTRACT_DATA_TOKEN,
+                .tuneId = TUNE_TAP_CASUAL
+            };
+            content->type = SWITCHES_LIST;
+            content->switchesList.nbSwitches = 3;
+            content->switchesList.switches = (nbgl_layoutSwitch_t*)switches;
+            break;
+        case 1:
+            content->type = INFOS_LIST;
+            content->infosList.nbInfos = 1;
+            content->infosList.infoTypes = (const char**)infoTypes;
+            content->infosList.infoContents = (const char**)infoContents;
+            break;
+    }
+    return true;
+}
+
+static void settings_ctrl(int token, uint8_t index) {
+    (void)index;
+    uint8_t value;
+
+    switch (token) {
+        case TESTNET_ADDR_TOKEN:
+            value = (N_data.address_type == ADDRESS_TESTNET) ? ADDRESS_MAINNET : ADDRESS_TESTNET;
+            nvm_write((void*)&N_data.address_type, (void*)&value, sizeof(value));
+            break;
+        case SIGN_HASH_TOKEN:
+            value = (N_data.sign_hash_type == SIGN_HASH_ON) ? SIGN_HASH_OFF : SIGN_HASH_ON;
+            nvm_write((void*)&N_data.sign_hash_type, (void*)&value, sizeof(value));
+            break;
+        case CONTRACT_DATA_TOKEN:
+            value = (N_data.contract_data_type == ALLOW_CONTRACT_DATA) ? DISALLOW_CONTRACT_DATA : ALLOW_CONTRACT_DATA;
+            nvm_write((void*)&N_data.contract_data_type, (void*)&value, sizeof(value));
+            break;
+    }
+}
+
+static void ui_settings(void) {
+    nbgl_useCaseSettings(APPNAME" settings",
+                         0,
+                         2,
+                         true,
+                         ui_initial_screen,
+                         settings_nav,
+                         settings_ctrl);
+}
+
+void ui_initial_screen(void) {
+    nbgl_useCaseHome(APPNAME,
+                     &C_stax_nervos_64px,
+                     NULL,
+                     true,
+                     ui_settings,
+                     exit_app);
+}
+
+static nbgl_layoutTagValue_t pair;
+static nbgl_layoutTagValueList_t pair_list;
+static nbgl_pageInfoLongPress_t info_long_press;
+char tag[MAX_SCREEN_COUNT][PROMPT_WIDTH + 1];
+char value[MAX_SCREEN_COUNT][VALUE_WIDTH + 1];
+
+static void onConfirmAbandon(void) {
+    nbgl_useCaseStatus("Transaction rejected", false, ui_initial_screen);
+    global.ui.cxl_callback();
+}
+
+static void reviewChoice(bool confirm) {
+    if (confirm) {
+        nbgl_useCaseStatus("TRANSACTION\nSIGNED", true, ui_initial_screen);
+        global.ui.ok_callback();
+    } else {
+        nbgl_useCaseConfirm("Reject transaction?",
+                            "",
+                            "Yes, Reject",
+                            "Go back to transaction",
+                            onConfirmAbandon);
+    }
+}
+
+static nbgl_layoutTagValue_t *get_pair(uint8_t index) {
+    global.ui.switch_screen(index);
```

Comment:
No checks required because it's done in switch_screen ? Maybe add a comment then.

Reply:
Done :+1: